dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java

```diff
@@ -0,0 +1,784 @@
+
+        if (name.startsWith("llvm.assume")) {
+            return List.of(EventFactory.newAssume(call.getArguments().get(0)));
+        } else if (name.startsWith("llvm.ctlz")) {
+            return inlineLLVMCtlz(valueCall);
+        } else if (name.startsWith("llvm.ctpop")) {
+            return inlineLLVMCtpop(valueCall);
+        } else if (name.contains("add.sat")) {
+            return inlineLLVMSaturatedAdd(valueCall);
+        } else if (name.contains("sub.sat")) {
+            return inlineLLVMSaturatedSub(valueCall);
+        } else if (name.startsWith("llvm.smax") || name.startsWith("llvm.smin")
+                || name.startsWith("llvm.umax") || name.startsWith("llvm.umin")) {
+            return inlineLLVMMinMax(valueCall);
+        } else {
+            final String error = String.format(
+                    "Call %s to LLVM intrinsic %s cannot be handled.", call, call.getCalledFunction());
+            throw new UnsupportedOperationException(error);
+        }
+    }
+
+    private List<Event> inlineLLVMCtlz(ValueFunctionCall call) {
+        final Expression input = call.getArguments().get(0);
+        // TODO: Handle the second parameter as well
+        final Register resultReg = call.getResultRegister();
+        final IntegerType type = (IntegerType) resultReg.getType();
+        final Expression increment = expressions.makeADD(resultReg, expressions.makeOne(type));
+
+        final List<Event> replacement = new ArrayList<>();
+        final Label endCtlz = EventFactory.newLabel("__ctlz_end");
+        replacement.add(EventFactory.newLocal(resultReg, expressions.makeZero(type)));
+        //TODO: There might be more efficient ways to count leading zeroes, though it is not clear
+        // if they are also more friendly for the SMT backend.
+        for (int i = type.getBitWidth() - 1; i >= 0; i--) {
+            final Expression testMask = expressions.makeValue(BigInteger.ONE.shiftLeft(i), type);
+            //TODO: dedicated test-bit expressions might yield better results, and they are supported by the SMT backend
+            // in the form of extract operations.
+            final Expression testBit = expressions.makeEQ(expressions.makeAND(input, testMask), testMask);
+
+            replacement.addAll(List.of(
+                    // FIXME: This generates a control dependency, but it should not!
+                    EventFactory.newJump(testBit, endCtlz),
+                    EventFactory.newLocal(resultReg, increment)
+                    // We could also do a direct assignment instead of an increment, but CP will resolve this either way.
+            ));
+        }
+        replacement.add(endCtlz);
+
+        return replacement;
+        /*final Expression result = expressions.makeCTLZ(input, (IntegerType) call.getResultRegister().getType());
+        return List.of(EventFactory.newLocal(call.getResultRegister(), result));*/
+    }
+
+    private List<Event> inlineLLVMCtpop(ValueFunctionCall call) {
```

Comment:
Can you describe what the intrinsic is suppose to do?

Reply:
It counts the number of bits that are 1. Do you want to have explanations for all llvm intrinsics added as comments?
Cause we could also just reference the LLVM manual

Reply:
The other at least have the implementation of the intrinsic that I can use to check if the implementation is correct. This one has nothing. A reference to the LLVM manual would be enough.

Reply:
What do you mean by the other one's have an implementation? All of them have an internal implementation.
You can also use `__builtin_ctpop` to generate the intrinsic in code and compare it against a manually implemented loop that counts the bits. I checked that this intrinsic works as expected on a few examples (I used the `ctlz.c` test and modified it).

Reply:
I meant "an implementation as documentation" like [here](https://github.com/hernanponcedeleon/Dat3M/blob/9d1b472bc597e5f2a1a469e38fd006418d9e6593/dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/Intrinsics.java#L445-L454)

Reply:
The implementation for the saturated operations is highly non-trivial, that's why I put it there. In particular, even if you know what saturated arithmetic does, it is not clear how to translate it properly. This is not the case for ctpop/ctlz, which are just straightforward bit-counting loops (i.e., once you know what they do, the implementation is trivial).
Anyway, I can write pseudo-code there, but then we should do this for all llvm intrinsics (at least for ctlz as well).

Reply:
If the implementation is trivial, there is no need as long as we provide a link to the documentation. This way we have a reference of what the implementation should do